### PR TITLE
New FK from campaigns to updates & migration to create updates from campaigns

### PR DESCRIPTION
**Pull request by houcros**
So we discussed that there should be a restriction so that each campaign must have an update object associated.
I created a new FK from campaigns to updates to enforce this. Then I had to refactor all the tests so that each time we create a campaign it has an existing update beforehand.
There is some code duplication which I still wasn't able to extract after much work, particularly the method `buildCampaignWithUpdate` is spread in a few classes. I wanted to extract it into a trait and mix it in but I didn't manage. So there is room for improvement.

### CODE REVIEW

File: src/main/scala/com/advancedtelematic/campaigner/db/Repository.scala [hunk 1]
```diff
@@ -213,7 +216,11 @@ protected class CampaignRepository()(implicit db: Database, ec: ExecutionContext
 
   def persist(campaign: Campaign, groups: Set[GroupId], metadata: Seq[CampaignMetadata]): Future[CampaignId] = db.run {
     val f = for {
-      _ <- (Schema.campaigns += campaign).handleIntegrityErrors(Errors.ConflictingCampaign)
+      _ <- (Schema.campaigns += campaign).asTry.flatMap {
```
Comment (houcros) [inline src/main/scala/com/advancedtelematic/campaigner/db/Repository.scala]: Also, is there a better way to do this? It feels hacky, but I wanted to differentiate and throw a different error when the cause is that there is no valid FK from campaigns to updates.

File: src/main/scala/com/advancedtelematic/campaigner/db/ExtractUpdatesFromCampaignsAndInsert.scala [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+  private val _log = LoggerFactory.getLogger(this.getClass)
+
+  private[db] def updateFromCampaign(c: Campaign): Update = {
+    val u = Update(UpdateId.generate(), UpdateSource(ExternalUpdateId(c.updateId.uuid.toString), UpdateType.multi_target), c.namespace, c.name, None, Instant.now, Instant.now)
+    _log.info("Extracted update {} from campaign {}", u.uuid, c.id, "")
+    u
+  }
+
+  def run(): Future[Done.type] = db.run {
```
[the remark this answers is not in the document]
Comment (houcros) [reply src/main/scala/com/advancedtelematic/campaigner/db/ExtractUpdatesFromCampaignsAndInsert.scala]: If there is no argument agains reusing `updateId` then I say yes, let's do it like this, much easier to read.
Isn't it `select c.update_id, c.update_id [...]` ?

[the remark this answers is not in the document]
Comment (simao) [reply src/main/scala/com/advancedtelematic/campaigner/db/ExtractUpdatesFromCampaignsAndInsert.scala]: @houcros it's equivalent but yes I'll change it

**Comment by simao** [comment]
@jerrytrieu  this is ready to go on my side

**Comment by simao** [comment]
@jerrytrieu  can we merge?

**Comment by jerrytrieu** [comment]
Still need to wait for UI to fully switch to new campaign/update.

**Comment by simao** [comment]
@houcros please check last commit

File: src/test/scala/com/advancedtelematic/campaigner/db/CampaignsSpec.scala [hunk 3]
```diff
@@ -105,7 +105,7 @@ class CampaignsSpec extends AsyncFlatSpec
   }
 
   "finishing devices" should "work with one campaign" in {
-    val group    = GroupId.generate()
+    val _    = GroupId.generate()
```
Comment (simao) [inline src/test/scala/com/advancedtelematic/campaigner/db/CampaignsSpec.scala]: then you don't need this line?

Comment (houcros) [reply src/test/scala/com/advancedtelematic/campaigner/db/CampaignsSpec.scala]: Right 😆 I removed it now.

**Pull request merged**